libraries/lib-wave-track/WaveTrack.cpp

```diff
@@ -358,6 +358,19 @@ void WaveTrack::DoOnProjectTempoChange(
          clip->OnProjectTempoChange(oldTempo, newTempo);
 }
 
+Finally<std::function<void()>>
+WaveTrack::ProvisionallySynchronizeProjectTempo(WaveTrack& other) const
+{
+   const auto originalTempo = other.GetProjectTempo();
+   const auto provisionalTempo = GetProjectTempo();
+   if (!provisionalTempo.has_value() || !originalTempo.has_value())
+      THROW_INCONSISTENCY_EXCEPTION;
+   other.OnProjectTempoChange(*provisionalTempo);
+   return { [originalTempo, &other] {
+      other.OnProjectTempoChange(*originalTempo);
```

Comment:
1. What happens inside a `Finally` may happen during stack unwinding, so the called functions should be `noexcept`.  (I don't believe in adding `noexcept` just because you can; but only when there are certain strong reasons for it like here.)

2. This might not restore the state of the other track exactly.  `mRawAudioTempo` in a clip may go from empty to non-empty and not back again.

Reply:
The two places that use this function also use `const_cast`.  I feel some unease.  I might approve reluctantly.

Reply:
>  the called functions should be `noexcept`

I never used that keyword before. I notice, though, that
```
void foo() { throw std::runtime_exception(""); }
void bar() noexcept { foo(); }
```
doesn't yield a warning, which would make tedious checking that all nodes of the call tree of `OnProjectTempoChange` also are `noexcept`. Or am I missing something ?

Reply:
> This might not restore the state of the other track exactly.

I hadn't thought about that, but looks like I'm a lucky guy: `track.GetProjectTempo().has_value()` if and only if `track.OnProjectTempoChange` was called at least once. If it wasn't, then this method doesn't proceed. If it was, then all clips have `mRawAudioTempo` set.